elastic/kibana
# [Security Solution][Endpoint] Adds logic for checking and completing Sentinelone `isolate` and `release` actions

## Summary

Adds checks for `isolate` and `release` response action for SentinelOne to see if they have completed. Changes include:

- `isolate` and `release` are no longer marked "complete" once request is sent to SentinelOne. They are kept in `pending` state
- Added logic to the SentinelOne Response Actions client to use the activity data from SentinelOne (index `logs-sentinel_one.activity-default`) and check if `isolate` and `release` action were complete
    - logic is invoked from the background task, which is not yet fully enabled
    - Note that the data in the SentinelOne index may be delayed. The frequency of how often that data is pulled in is defined via the SentinelOne Integration
- Enhanced the SentinelOne host script so that it also automatically generates a SentinelOne alert when the host VM running the SentinelOne agent is created
- Refactor: Moved`server/endpoint/mocks.ts` file to its own directory (`server/endpoint/mocks/**`)
    - File is getting to large... having it in a directory will allows us to break this file down into smaller pieces while still exporting out of `server/endpoint/mocks` path.
    - Added new `utils.mock.ts` that includes 1 new generic utilty - `applyEsClientSearchMock()`



<details><summary>Sample output to kibana log (when running with `debug` mode)</summary>

### Output showing completion of an Isolate action by checking SentinelOne's activity log:

```
[2024-04-05T08:50:08.357-04:00][DEBUG][plugins.securitySolution.CompleteExternalActionsTaskRunner.c29i02] Started: Checking status of external response actions
[2024-04-05T08:50:08.365-04:00][DEBUG][plugins.securitySolution.SentinelOneActionsClient] searching for isolate responses from [logs-sentinel_one.activity-default] index with:
{
  index: 'logs-sentinel_one.activity-default',
  query: {
    bool: {
      must: [
        { terms: { 'sentinel_one.activity.type': [ 1001, 2010 ] } }
      ],
      should: [
        {
          bool: {
            filter: [
              {
                term: {
                  'sentinel_one.activity.agent.id': '1913920934584665209'
                }
              },
              {
                range: {
                  'sentinel_one.activity.updated_at': { gt: '2024-04-05T12:49:12.113Z' }
                }
              }
            ]
          }
        }
      ],
      minimum_should_match: 1
    }
  },
  collapse: {
    field: 'sentinel_one.activity.agent.id',
    inner_hits: {
      name: 'first_found',
      size: 1,
      sort: [ { 'sentinel_one.activity.updated_at': 'asc' } ]
    }
  },
  _source: false,
  sort: [ { 'sentinel_one.activity.updated_at': { order: 'asc' } } ],
  size: 1000
}
[2024-04-05T08:50:08.421-04:00][DEBUG][plugins.securitySolution.SentinelOneActionsClient] Search results for SentinelOne isolate activity documents:
{
  took: 46,
  timed_out: false,
  _shards: { total: 1, successful: 1, skipped: 0, failed: 0 },
  hits: {
    total: { value: 1, relation: 'eq' },
    max_score: null,
    hits: [
      {
        _index: '.ds-logs-sentinel_one.activity-default-2024.04.05-000001',
        _id: 'UP8uvXmyc0P6OBys00iGA2gzFOc=',
        _score: null,
        fields: { 'sentinel_one.activity.agent.id': [ '1913920934584665209' ] },
        sort: [ 1712321353320 ],
        inner_hits: {
          first_found: {
            hits: {
              total: { value: 1, relation: 'eq' },
              max_score: null,
              hits: [
                {
                  _index: '.ds-logs-sentinel_one.activity-default-2024.04.05-000001',
                  _id: 'UP8uvXmyc0P6OBys00iGA2gzFOc=',
                  _score: null,
                  _source: [Object],
                  sort: [Array]
                }
              ]
            }
          }
        }
      }
    ]
  }
}
[2024-04-05T08:50:08.422-04:00][DEBUG][plugins.securitySolution.SentinelOneActionsClient] 1 isolate action responses generated:
[
  {
    '@timestamp': '2024-04-05T12:50:08.422Z',
    agent: { id: 'c06d63d9-9fa2-046d-e91e-dc94cf6695d8' },
    EndpointActions: {
      action_id: '583b8a6c-0227-497e-9bbf-8af551e69ed6',
      input_type: 'sentinel_one',
      started_at: '2024-04-05T12:50:08.422Z',
      completed_at: '2024-04-05T12:50:08.422Z',
      data: { command: 'isolate' }
    },
    error: undefined,
    meta: {
      elasticDocId: 'UP8uvXmyc0P6OBys00iGA2gzFOc=',
      activityLogEntryId: '1921767625558733652',
      activityLogEntryType: 1001,
      activityLogEntryDescription: 'Agent ptavares-sentinelone-1371 was disconnected from network.'
    }
  }
]
[2024-04-05T08:50:08.426-04:00][DEBUG][plugins.securitySolution.CompleteExternalActionsTaskRunner.c29i02.QueueProcessor] Processing batch [1] with [1] items. Items remaining in queue: [0]
[2024-04-05T08:50:09.515-04:00][DEBUG][plugins.securitySolution.CompleteExternalActionsTaskRunner.c29i02.QueueProcessor] Processed [1] batches and a total of [1] items

```


</details>




### How to Test

- Bring up the stack
- Setup SentinelOne host. you can can the following script to get it all setup: `node x-pack/plugins/security_solution/scripts/endpoint/run_sentinelone_host.js` . The credentials needed to setup the host can be retrieved from [here](https://p.elstc.co/paste/2TIZH79o#hDuLXQ9+h8DKk1nghayfPQIdkJcbyEXYVRJ5sAzlQOz).
- `isolate` or `release` the host. The response actions should remain in `pending` until the action is complete in SentinelOne and data (sentinelone activity logs) is ingested back to elasticsearch. NOTE: this could take a bit of time depending on the frequency setup in the integration policy for sentinelone



### Checklist
